File: python/test_modifying_sequence.py
import pytest
from modifying_sequence import *

@pytest.fixture
def ary():
    return [1, 2, 3, 4, 5]

@pytest.fixture
def set_x():
    return set([1, 2, 3])

@pytest.fixture
def dict_x():
    return dict([(1, '1'), (2, '2')])

def test_copy(ary):
    assert [i for i in copy(ary, 0, 3)] == ary[:3]

def test_copy_if(ary):
    assert [i for i in copy_if(ary, lambda x: x < 4)] == [1, 2, 3]

def test_copy_n(ary):
    assert [i for i in copy_n(ary, 3)] == ary[:3]

def test_copy_backward(ary):
    assert [i for i in copy_backward(ary, 0, 3)] == ary[2::-1]

def test_move(ary):
    assert [i for i in move(ary, 0, 3)] == ary[:3]

def test_move_backward(ary):
    assert [i for i in move_backward(ary, 0, 3)] == ary[2::-1]

def test_fill(ary):
    fill(ary, 3)
    assert ary == [3, 3, 3, 3, 3]

def test_fill_n(ary):
    fill_n(ary, 3, 2)
    assert ary == [3, 3, 3, 4, 5]

def test_transform(ary, dict_x):
    a = transform(ary, lambda x: x + 1)
    assert a == [2, 3, 4, 5, 6]
    b = dict(transform(dict_x, lambda k: (k, k)))
    assert b == dict([(1, 1), (2, 2)])

def test_generate(ary):
    generate(ary, lambda :3)
    assert ary == [3, 3, 3, 3, 3]

def test_generate_n(ary):
    generate_n(ary, lambda :3, 3)
    assert ary == [3, 3, 3, 4, 5]

def test_remove(ary, set_x):
    assert [i for i in remove(ary, 4)] == [1, 2, 3, 5]
    assert set([i for i in remove(set_x, 3)]) == set([1, 2])
    # can't work on dict
    # assert [i for i in remove(dict_x, 1)] == [(2, '2')]

def test_remove_if(ary):
    assert [i for i in remove_if(ary, lambda x: x < 3)] == [3, 4, 5]

def test_remove_copy(ary, set_x):
    a = remove_copy(ary, 4)
    assert a == [1, 2, 3, 5]
    b = set(remove_copy(set_x, 3))
    assert b == set([1, 2])

def test_remove_copy_if(ary):
    a = remove_copy_if(ary, lambda x: x < 3)
    assert a == [3, 4, 5]

def test_replace(ary, set_x):
    assert [i for i in replace(ary, 4, 5)] == [1, 2, 3, 5, 5]
    assert set([i for i in replace(set_x, 3, 4)]) == set([1, 2, 4])

def test_replace_if(ary):
    assert [i for i in replace_if(ary, lambda x: x < 3, 1)] == \
            [1, 1, 3, 4, 5]

def test_replace_copy(ary, set_x):
    a = replace_copy(ary, 4, 3)
    assert a == [1, 2, 3, 3, 5]
    b = set(replace_copy(set_x, 3, 4))
    assert b == set([1, 2, 4])

def test_replace_copy_if(ary):
    a = replace_copy_if(ary, lambda x: x < 3, 1)
    assert a == [1, 1, 3, 4, 5]

File: ruby/test_modifying_sequence.rb
require 'minitest/autorun'
require 'set'
require_relative 'modifying_sequence'

class TestModifyingSequence < MiniTest::Test
  include ModifyingSequence

  def setup
    @ary = [1, 2, 3, 4, 5]
    @set = Set.new([1, 2, 3])
    @hash = {1 => '1', 2 => '2'}
    @duplicate_ary = [1, 2, 2, 2, 4]
  end

  def test_copy
    assert_equal @ary[0..2], copy(@ary, 0, 3)
  end

  def test_copy_if
    assert_equal [1, 2, 3], copy_if(@ary){|x| x < 4}
  end

  def test_copy_n
    assert_equal @ary[0..2], copy_n(@ary, 3)
  end

  def test_copy_backward
    assert_equal @ary[0..2].reverse, copy_backward(@ary, 0, 3)
  end

  # move is as the same as copy
  def test_move
    assert_equal @ary[0..2], move(@ary, 0, 3)
  end

  def test_move_backward
    assert_equal @ary[0..2].reverse, move_backward(@ary, 0, 3)
  end

  def test_fill
    fill(@ary, 3)
    assert_equal Array.new(5, 3), @ary
  end

  def test_transform
    assert_equal [2, 3, 4, 5, 6], transform(@ary){|x| x + 1}
  end

  def test_generate
    assert_equal Array.new(5, 3), generate(@ary){3}
  end

  def test_generate_n
    assert_equal [3, 3, 3, 4, 5], generate_n(@ary, 3){3}
  end

  def test_remove
    assert_equal [1, 2, 3, 5], remove(@ary, 4)
    assert_equal Set.new([1, 2]), Set.new(remove(@set, 3))
    removed_hash = {2 => '2'}
    assert_equal removed_hash, remove(@hash, 1)
  end

  def test_remove_if
    assert_equal [3, 4, 5], remove_if(@ary){|x| x < 3}
  end

  def test_remove_copy
    assert_equal [1, 2, 3, 5], remove_copy(@ary, 4)
    assert_equal Set.new([1, 2]), Set.new(remove_copy(@set, 3))
  end

  def test_remove_copy_if
    assert_equal [3, 4, 5], remove_copy_if(@ary){|x| x < 3}
  end

  def test_replace
    assert_equal [1, 2, 3, 5, 5], replace(@ary, 4, 5)
    assert_equal Set.new([1, 2, 4]), Set.new(replace(@set, 3, 4))
  end

  def test_replace_if
    assert_equal [1, 1, 3, 4, 5], replace_if(@ary, 1){|x| x < 3}
  end

  def test_swap
    another_ary = [5, 4, 3, 2, 1]
    after_swap_ary = Array.new(another_ary)
    after_swap_another_ary = Array.new(@ary)
    @ary, another_ary = swap(@ary, another_ary)
    assert_equal after_swap_ary, @ary
    assert_equal after_swap_another_ary, another_ary
  end

  def test_swap_range
    another_ary = [5, 4, 3]
    swap_range(@ary, another_ary, 0, 3)
    assert_equal [5, 4, 3, 4, 5], @ary
    assert_equal [1, 2, 3], another_ary
  end

  def test_reverse
    reverse(@ary)
    assert_equal [5, 4, 3, 2, 1], @ary
  end

  def test_rotate
    assert_equal [4, 5, 1, 2, 3], rotate(@ary, 3)
  end

  def test_rotate_copy
    assert_equal [4, 5, 1, 2, 3], rotate_copy(@ary, 3)
  end

  def test_shuffle
    shuffle @ary
  end

  def test_unique
    assert_equal [1, 2, 4], unique(@duplicate_ary)
  end
end
